src/core/llm.py

```diff
@@ -0,0 +1,89 @@
+from enum import StrEnum, auto
+from functools import cache
+from typing import TypeAlias
+
+from langchain_anthropic import ChatAnthropic
+from langchain_aws import ChatBedrock
+from langchain_google_genai import ChatGoogleGenerativeAI
+from langchain_groq import ChatGroq
+from langchain_openai import ChatOpenAI
+
+
+class Provider(StrEnum):
+    OPENAI = auto()
+    ANTHROPIC = auto()
+    GOOGLE = auto()
+    GROQ = auto()
+    AWS = auto()
+
+
+class OpenAIModelName(StrEnum):
+    """https://platform.openai.com/docs/models/gpt-4o"""
+
+    GPT_4O_MINI = "gpt-4o-mini"
+    GPT_4O = "gpt-4o"
+
+
+class AnthropicModelName(StrEnum):
+    """https://docs.anthropic.com/en/docs/about-claude/models#model-names"""
+
+    HAIKU_3 = "claude-3-haiku"
+    HAIKU_35 = "claude-3.5-haiku"
+    SONNET_35 = "claude-3.5-sonnet"
+
+
+class GoogleModelName(StrEnum):
+    GEMINI_15_FLASH = "gemini-1.5-flash"
+
+
+class GroqModelName(StrEnum):
+    """https://console.groq.com/docs/models"""
+
+    LLAMA_31_8B = "groq-llama-3.1-8b"
+    LLAMA_31_70B = "groq-llama-3.1-70b"
+
+    LlAMA_GUARD_3_8B = "groq-llama-guard-3-8b"
+
+
+class AWSModelName(StrEnum):
+    BEDROCK_HAIKU = "bedrock-3.5-haiku"
+
+
+_MODEL_TABLE = {
+    OpenAIModelName.GPT_4O_MINI: "gpt-4o-mini",
+    OpenAIModelName.GPT_4O: "gpt-4o",
+    AnthropicModelName.HAIKU_3: "claude-3-haiku-20240307",
+    AnthropicModelName.HAIKU_35: "claude-3-5-haiku-latest",
+    AnthropicModelName.SONNET_35: "claude-3-5-sonnet-latest",
+    GoogleModelName.GEMINI_15_FLASH: "gemini-1.5-flash",
+    GroqModelName.LLAMA_31_8B: "llama-3.1-8b-instant",
+    GroqModelName.LLAMA_31_70B: "llama-3.1-70b-versatile",
+    GroqModelName.LlAMA_GUARD_3_8B: "llama-guard-3-8b",
+    AWSModelName.BEDROCK_HAIKU: "anthropic.claude-3-5-haiku-20241022-v1:0",
+}
+
+AllModelEnum: TypeAlias = (
+    OpenAIModelName | AnthropicModelName | GoogleModelName | GroqModelName | AWSModelName
+)
+ModelT: TypeAlias = ChatOpenAI | ChatAnthropic | ChatGoogleGenerativeAI | ChatGroq | ChatBedrock
+
+
+@cache
```

Comment:
Is this optimization doing anything useful here? I'm slightly nervous about side effects or weirdness for negligible benefit... maybe it's fine though

Reply:
Oh sorry I was referring to the use of @cache

Reply:
I used the @cache method to ensure that when the same model_name is input, the same object can be retrieved without having to recreate it.
In fact, these BaseChatModels seem to require creating an httpx.client internally, which makes object creation take quite a bit of time. On a better computer, it might take only 0.05 to 0.3 seconds, while on a less powerful one, it could take even longer.

Essentially, this cache is just an alternative to using a global dict to store models. I think that writing it as a function is easier to read than directly accessing objects from a dict using a key.

Reply:
Sounds good, I didn't realize about the httpx client. Good catch! 👍